scvi/inference/posterior.py

```diff
@@ -505,39 +506,60 @@ def imputation(self, n_samples=1):
         return imputed_list.squeeze()
 
     @torch.no_grad()
-    def generate(self, n_samples=100, genes=None):  # with n_samples>1 return original list/ otherwose sequential
+    def generate(
+        self,
+        n_samples: int = 100,
+        genes: Union[list, np.ndarray] = None,
+        batch_size: int = 128
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
         """
-        Return original_values as y and generated as x (for posterior density visualization)
-        :param n_samples:
-        :param genes:
-        :return:
+        Create observation samples from the Posterior Predictive distribution
+
+        :param n_samples: Number of required samples for each cell
+        :param genes: Indices of genes of interest
+        :param batch_size: Desired Batch size to generate data
+
+        :return: Tuple (x_new, x_old)
+            Where x_old has shape (n_cells, n_genes)
+            Where x_new has shape (n_cells, n_genes, n_samples)
         """
-        original_list = []
-        posterior_list = []
-        batch_size = 128  # max(self.data_loader_kwargs["batch_size"] // n_samples, 2)  # Reduce batch_size on GPU
+        assert self.model.reconstruction_loss in ['zinb', 'nb']
+        zero_inflated = self.model.reconstruction_loss == 'zinb'
+        x_old = []
+        x_new = []
         for tensors in self.update({"batch_size": batch_size}):
```

Comment:
Why exactly is this necessary?  Why can't we iterate through self?

Reply:
``for sample_batch, _, _, batch_index, labels in self.update({"batch_size": batch_size}):`` is more concise indeed.

Reply:
@gabmis I guess I was asking why we need to update the batch size? Doesn't that make a copy? If the tensors get put on GPU during every iteration, why do we need to change the batch size?

Reply:
I am not sure I get your inquiry Adam. IMO, ``batch_size`` seems a like a reasonable parameter for this function. Besides, making a shallow copy of ``Posterior`` object should be rather inexpensive.